File: freight-triage-polyglot-ledger-recovery/environment/reconcile/internal/statx/ewma_milli.go
package statx

// EwmaMilli is the ewma_milli kernel.
func EwmaMilli(series []int64) int64 {
	if len(series) == 0 {
		return 0
	}
	state := series[0] * 1000
	for _, value := range series[1:] {
		state += FloorDiv(value*1000-state, 4)
	}
	return state
}

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/stats/EwmaMilli.java
package com.freight.stats;

/** ewma_milli kernel. */
public final class EwmaMilli implements StatKernel {

    @Override
    public String name() {
        return "ewma_milli";
    }

    @Override
    public long apply(long[] series) {
        if (series.length == 0) {
            return 0L;
        }
        long state = series[0] * 1000L;
        for (int i = 1; i < series.length; i++) {
            state += StatSupport.floorDiv(series[i] * 1000L - state, 4L);
        }
        return state;
    }
}
